Release/src/http/client/http_client_asio.cpp

```diff
@@ -350,22 +370,374 @@ class asio_client : public _http_client_communicator, public std::enable_shared_
+    
+    std::string toString(void *id, tcp::resolver::iterator endpoints)
+    {
+        std::stringstream ss;
+        
+        auto host = endpoints->host_name();
+        auto service = endpoints->service_name();
+        auto address = endpoints->endpoint().address();
+        auto ip = address.to_string();
+        
+        ss << "id: " << id <<  toString(endpoints);
+        
+        return ss.str();
+    }
+    
+    void removePendingConnection(std::shared_ptr<asio_connection> connection)
+    {
+        pendingConnections.erase(std::remove(pendingConnections.begin(), pendingConnections.end(), connection), pendingConnections.end());
+    }
+    
+    void handle_tcp_connect(std::weak_ptr<asio_connection> weakConnection, int id, const boost::system::error_code& ec, tcp::resolver::iterator endpoints, ConnectHandler handler)
+    {
+        auto connection = weakConnection.lock();
+        if (!connection)
+        {
+            return;
+        }
+        
+        if (!ec)
+        {
+            switch (testAndSetState(ConnectionState::ConnectedSuccess))
+            {
+                case ConnectionState::ConnectedFailed:
+                case ConnectionState::Connecting:
+                    {
+                        ScopedLock l(m_lock);
+                        
+                        LOG("cpprest: Connected to: host: " << toString(connection.get(), endpoints));
+                        
+                        cancelHeTimer();
+                    
+                        // Cache succesful endpoint only when happy eyeballs was used
+                        if (pendingConnections.size() > 1)
+                        {
+                            LOG("cpprest: Added to cache host: " << toString(connection.get(), endpoints));
+                            getCache()->add({endpoints->host_name(), endpoints->service_name()}, endpoints->endpoint());
+                        }
+                        pendingConnections.clear();
+                    }
+                    handler(connection, ec, endpoints);
+                    break;
+                case ConnectionState::Idle:
+                    // Invalid state at that point
+                    {
+                        ScopedLock l(m_lock);
+                        removePendingConnection(connection);
+                    }
+                    assert(true);
+                    break;
+                case ConnectionState::ConnectedSuccess:
+                    // It will auto disconnect after we exit this function
+                    {
+                        ScopedLock l(m_lock);
+                        removePendingConnection(connection);
+                    }
+                    LOG("cpprest: Already connected: host: " << toString(connection.get(), endpoints));
+                    break;
+            }
+        }
+        else if (++endpoints == tcp::resolver::iterator())
+        {
+            switch (testAndSetState(ConnectionState::ConnectedFailed))
```

Comment:
in case the last possible endpoint was hit and returned error:

1. remove the connection from pending connections.